### PR TITLE
Verify entered homeserver before proxy availability.

### PR DESCRIPTION
**Pull request by pixlwave**
Passing a URL directly to the AuthenticationService would always fail as the well-known discovery doesn't take place. But it also doesn't report back to the user whether or not there is a valid homeserver at that URL, merely saying sliding sync isn't available which isn't helpful. This PR fixes https://github.com/vector-im/element-x-ios/issues/591 in 2 ways:
- When a URL is entered attempt discovery with a server name first before falling back to the URL directly.
- Check for homeserver details before the proxy so that entering "https://idontexist.com/" fails with the correct error.

### CODE REVIEW

**Review by jplatte — COMMENTED** [review]
I'm not happy with extending the amount of logic that's in the FFI layer. If we want to provide HS sanitization, it should be a public function in the `matrix-sdk` crate, something like

```rust
fn sanitize_server_name(s: &str) -> Result<OwnedServerName, SomeError>
```

**Comment by pixlwave** [comment]
> I'm not happy with extending the amount of logic that's in the FFI layer. If we want to provide HS sanitization, it should be a public function in the `matrix-sdk` crate, something like
> 
> ```rust
> fn sanitize_server_name(s: &str) -> Result<OwnedServerName, SomeError>
> ```

**Comment by pixlwave** [comment]
@jplatte How does that look? I just put the function in `lib.rs`, happy to move it to a more suitable place though :)

File: bindings/matrix-sdk-ffi/src/authentication_service.rs [hunk 1]
```diff
@@ -112,32 +120,48 @@ impl AuthenticationService {
...
+                // When discovery fails, fallback to the homeserver URL if supplied.
+                let mut builder = Arc::new(ClientBuilder::new()).base_path(self.base_path.clone());
+                builder = builder.homeserver_url(server_name_or_homeserver_url);
+                builder.build()
+            })
+            .map_err(AuthenticationError::from)?;
 
-        let client = builder.build().map_err(AuthenticationError::from)?;
+        let details =
+            RUNTIME.block_on(async /*move*/ { self.details_from_client(&client).await })?;
```
Comment (pixlwave) [inline bindings/matrix-sdk-ffi/src/authentication_service.rs:151]: The commented `move` obviously need to be removed from here. I left it to ask the question that this was safe to remove from the implementation. Assumed yes given the compile was happy.

**Review by jplatte — APPROVED** [review]

File: bindings/matrix-sdk-ffi/src/authentication_service.rs (hunk 1 above)
Comment (jplatte) [reply bindings/matrix-sdk-ffi/src/authentication_service.rs]: As long as it compiles, `move` or no `move` hardly matters.

**Pull request merged**